### PR TITLE
Opt-out switch for DynamicResource optimization

### PR DESCRIPTION
**Pull request by dipeshmsft**
## Description
Implements the opt-out switch for DynamicResource optimization implemented in ( #5610, #9393 )

## Customer Impact
See here https://github.com/dotnet/wpf/pull/5610#issuecomment-2262959543

## Regression
N/A

## Testing
Local Testing with sample apps

## Risk

cc @miloush @batzen @h3xds1nz 

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/wpf/pull/9501)

### CODE REVIEW

File: src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/ResourceDictionary.cs [hunk 1]
```diff
@@ -1735,20 +1735,43 @@ private void NotifyOwners(ResourcesChangeInfo info)
                     DeferredResourceReference deferredResourceReference;
                     if (!IsThemeDictionary)
                     {
-                        // Cache the deferredResourceReference so that it can be validated
-                        // in case of a dictionary change prior to its inflation
-                        _deferredResourceReferences ??= new DeferredResourceReferenceList();
-
-                        if (_deferredResourceReferences.Get(resourceKey) is { } existingDeferredResourceReference
-                            && existingDeferredResourceReference.Dictionary == this)
+                        if(FrameworkAppContextSwitches.DisableDynamicResourceOptimization)
```
Comment (h3xds1nz) [inline src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/ResourceDictionary.cs:1738]: `if(FrameworkAppContextSwitches.DisableDynamicResourceOptimization)`

Can we just make sure all 6 occurrences have space after if?

File: src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/ResourceDictionary.cs [hunk 2]
```diff
@@ -1735,20 +1735,43 @@ private void NotifyOwners(ResourcesChangeInfo info)
...
+                            }
+
+                            // Cache the deferredResourceReference so that it can be validated
+                            // in case of a dictionary change prior to its inflation
+                            if (_weakDeferredResourceReferences == null)
+                            {
+                                _weakDeferredResourceReferences = new WeakReferenceList();
+                            }
+                            
+                            _weakDeferredResourceReferences.Add( deferredResourceReference, true /*SkipFind*/);
```
Comment (h3xds1nz) [inline src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/ResourceDictionary.cs:1756]: Remove space after `(`, guess old code.

File: src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/ResourceDictionary.cs [hunk 3]
```diff
@@ -1735,20 +1735,43 @@ private void NotifyOwners(ResourcesChangeInfo info)
...
-                        // Cache the deferredResourceReference so that it can be validated
-                        // in case of a dictionary change prior to its inflation
-                        _deferredResourceReferences ??= new DeferredResourceReferenceList();
-
-                        if (_deferredResourceReferences.Get(resourceKey) is { } existingDeferredResourceReference
-                            && existingDeferredResourceReference.Dictionary == this)
+                        if(FrameworkAppContextSwitches.DisableDynamicResourceOptimization)
                         {
-                            deferredResourceReference = existingDeferredResourceReference;
+                            if (_ownerApps != null)
```
Comment (RussKie) [inline src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/ResourceDictionary.cs:1740]: [nit]
```suggestion
                            if (_ownerApps is not null)
```

File: src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/ResourceDictionary.cs [hunk 4]
```diff
@@ -1735,20 +1735,43 @@ private void NotifyOwners(ResourcesChangeInfo info)
...
+                                deferredResourceReference = new DeferredAppResourceReference(this, resourceKey);
+                            }
+                            else
+                            {
+                                deferredResourceReference = new DeferredResourceReference(this, resourceKey);
+                            }
+
+                            // Cache the deferredResourceReference so that it can be validated
+                            // in case of a dictionary change prior to its inflation
+                            if (_weakDeferredResourceReferences == null)
```
Comment (RussKie) [inline src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/ResourceDictionary.cs:1751]: [nit]
```suggestion
                            if (_weakDeferredResourceReferences is null)
```

File: src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/ResourceDictionary.cs (hunk 1 above)
Comment (RussKie) [reply src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/ResourceDictionary.cs:1738]: There are incorrect spacing issues in this changeset. I recommend running "format document" in VS
image

**Comment by miloush** [comment]
Out of curiosity, if we want to just ignore the exception, why don't we go with `try catch` instead of a switch?

File: src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/ResourceDictionary.cs (hunk 1 above)
Comment (dipeshmsft) [reply src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/ResourceDictionary.cs:1738]: Sure, I will go ahead and do this.
I will try format document, but I am not very keen on it, if it does a lot of formatting.

Comment (dipeshmsft) [reply src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/ResourceDictionary.cs]: Fixed this

File: src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/ResourceDictionary.cs (hunk 3 above)
Comment (dipeshmsft) [reply src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/ResourceDictionary.cs]: Fixed this.

File: src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/ResourceDictionary.cs (hunk 4 above)
Comment (dipeshmsft) [reply src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/ResourceDictionary.cs]: Fixed this.

File: src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/ResourceDictionary.cs (hunk 2 above)
Comment (dipeshmsft) [reply src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/ResourceDictionary.cs]: Fixed this.

Comment (h3xds1nz) [reply src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/ResourceDictionary.cs]: LGTM

**Comment by dipeshmsft** [comment]
@miloush, there were a few reasons for that- first, where to add the try-catch block, the only place where it seemed plausible was ResourceReferenceExpression and the exception occurs on Invalidation method calls so catching the exception there would result in skipping invalidations. Moreover, handling XamlParseException may lead to swallowing other valid exceptions.

File: src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/ResourceDictionary.cs [hunk 5]
```diff
@@ -1771,35 +1794,55 @@ private void NotifyOwners(ResourcesChangeInfo info)
...
-            return;
+                    Inflate(deferredResourceReference);
+                }
 
-            void Inflate(DeferredResourceReference deferredResourceReference)
-            {
-                if (deferredResourceReference is not null)
+                return;
+
+                void Inflate(DeferredResourceReference deferredResourceReference)
```
Comment (hiteshwpfmsft) [inline src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/ResourceDictionary.cs:1837]: nit - Indentation for this method seems off. could you please fix it.

**Review by hiteshwpfmsft — APPROVED** [review]
LGTM

File: src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/ResourceDictionary.cs (hunk 5 above)
Comment (dipeshmsft) [reply src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/ResourceDictionary.cs:1837]: It seems off, but it is correct

**Pull request merged**